syfertext/utils.py

```diff
@@ -131,9 +131,62 @@ def _download_model(model_name: str):
                     # Downloading the file finished
                     break
     prog_bar.close()
+
     return tmp_model_path
 
 
+# The following function for normalizing is inspired
+# from Spacy  https://github.com/explosion/spaCy/blob/master/spacy/util.py.
+def normalize_slice(length: int, start: int, stop: int, step: int = None):
+    """This function is used to convert the negative slice boundaries to positive values.
+    eg. start = -4, stop = -1, length = 6 gets converted to start = 2, stop = 5
+
+    Args:
+        length (int): the length of the document to slice
+        start (int): the start index of the slice
+        stop (int): the stop index of the slice
+        step (int): the step value for the slice
+
+    Returns:
+        (start, stop) : pair of non-negative integer values signifying the
+                        normalized values of the slice
+    """
+    assert step is None or step == 1, "Stepped slices with steps greater than one are not supported"
+
+    # if start is none, that means we need to start from 0 index
+    if start is None:
+        start = 0
+
+    # if start is negative, we add the length to get its actual index
+    elif start < 0:
+        start += length
+
+    # start should not exceed the length of the document
+    # also max(0,start) ensures the start is never negative
+    start = min(length, max(0, start))
+
+    # stop is None, that means we need stop to be the last index+1
+    if stop is None:
+        stop = length
+
+    # add the length to get the actual positive index for stop if
+    # is negative
+    elif stop < 0:
+        stop += length
+
+    # stop should be less than or equal to length. Also max(start,stop) ensures that start <= stop
+    stop = min(length, max(start, stop))
+
+    # if this occurs, that means we have an empty range.
+    if start == stop:
```

Comment:
Actually `start > stop` never happens because of the statement `stop = min(length, max(start, stop))`. Instead when we have an empty range, `start = stop` in that case.

Reply:
what if start > length?

Reply:
`start <= length` always because of  the line 53 ie. `start = min(length, max(0, start))`. So whenever we give an invalid or empty range, these assignments of `start` and `stop` always lead them to be equal.
So invarient of `start` is that `0 <= start <= length` and for stop is `0 <= start <= stop <= length`.
eg. say length of `doc` is `10`. and we do `doc[1 : 12]`, then `start = 1` and `stop = 10` 
if we do `doc[5 :  3]`, `start = 5 , stop = 5`
for `doc[13 : 15]`, `start = 10, stop = 10`
That is why only thing we need to check for invalid or empty range is that if `start == stop`

Reply:
You are right, then no need for the condition `if start == stop` it does not change the return value, right?

Reply:
I didn't get the "change in return value" part ? Could you elaborate ? 😅

Reply:
I mean if `start == stop` when range is invalid or empty, then returning `(5,5)` or `(10,10)` doesn't look neat. So I just return `(0,0)` when this happens. And yes it wont make any difference if we remove the statement `if start == stop`.

Reply:
Well I still suggest to remove it. This will remove an unnecessary operation. save some time. even miniscule. The use do not have access to the returned values. So no problem in my opinion of it not being neat.